Playback page, Database, Recording page bug fixes

Alec-Hallman opened the pull request:
This PR features:
- Implementation of the data base
- Playback page
- the homepage now updates when a new recording is made. Displaying the mist recent recording without needing to restart the whole app. 
- Turned recording page which won't save any files to the data base that are less than 2 seconds in length. I also fixed the permissions button issue as well as streamlined a boolean variable which wasn't really needed.
- Cranked up the bitrate and the sample rate to create crispy ah hell audio. Recordings no longer sound like they were captured on a 1990 Blackberry phone.
- NO API YET (ill get there hopefully b4 tuesday)


> Play back page:

![image](https://github.com/LaurierCS/POD1/assets/109433535/d54de817-370b-4c4d-b2da-8356dfd4eb4e)

> Home page with recent recordings list

![image](https://github.com/LaurierCS/POD1/assets/109433535/f7012711-4b6e-4e47-912b-77890711ad91)

Alec-Hallman committed:
Playback page, Database, Recording page bug fixes

Alec-Hallman committed:
Recording bug fix

I had some insanely silly wacky zany code that only ran the first time the recording page was opened. Including the name of the file. So files were over writing themselves. Fixed that though, by moving some code around.

pabloj2001 commented on mobile/voice_journal_app/lib/Playback.dart:
```diff
@@ -0,0 +1,157 @@
+class _PlaybackPageState extends State<PlaybackPage>{
+  changeState(){
+    if(!listening){ //if this is the first time the button is being pressed
+      controller.onCompletion.listen((event){ pauseRecording(); setState((){});}); //add a listener so that when the file ends it pauses. This allows the user to loop their recording as many times as they want, but they just have to hit the play button to do so
```
Let's set this up when we first create the controller so we don't have to check every time the state changes. So let's put it in the `DisplayRecording` function.

pabloj2001 commented on mobile/voice_journal_app/lib/Playback.dart:
```diff
@@ -0,0 +1,157 @@
+int durration = 0;
+//----Done Initializing variables----
+
+DisplayRecording(Recording givenRecording) async{ //function to get information out of the recording being shown
```
Let's make this function name `displayRecording` so it's clear that it's a function.

pabloj2001 commented on mobile/voice_journal_app/lib/Playback.dart:
```diff
@@ -0,0 +1,157 @@
+}
+void main(){
+  runApp(const PlaybackPage(title: 'playback Page'));
+}
```
We should only have one `main` function in our entire app, let's remove this since this page won't be the entry point for our app.

pabloj2001 commented on mobile/voice_journal_app/lib/RecordingPage.dart:
```diff
@@ -35,30 +46,38 @@ void displaySaved(){ //Create a little pop up letting the user know their reccor
+  recording = true; 
+  //rbox = await Hive.openBox<Recording>('recordings'); //open the box so we can put things inside of it.
+  rbox = Hive.box<Recording>('recordings');
+  currentRecording = Recording(id: 'api call',title: formattedDateTime,audioFile: path,transcriptFile: 'api call',timeStamp: presently,emotion: Emotions.happy,isTranscribed: false,transcriptionId:'api call', durration: 0); //create the recording class
```
The emotions field should be an array of emotions so we can have more than one per recording. It should also be initialized as empty for a new recording. Likewise, `transcriptionId` will be null or empty since we have not yet recorded the entry so we haven't called the api. `duration` has a small spelling error.

pabloj2001 commented on mobile/voice_journal_app/lib/RecordingPage.dart:
```diff
@@ -67,27 +86,36 @@ pauseRecording(){
 }
+void main() async{
+  runApp(const MaterialApp(home: HomePage())); //Run the homepage if this page is ran on its own
+}
```
Not needed

pabloj2001 replied:
we don't need this main func here

pabloj2001 commented on mobile/voice_journal_app/lib/main.dart:
```diff
@@ -1,17 +1,20 @@
+void main()async { 
+  await Hive.initFlutter(); //Initialize hive for flutter crucial step
+  Hive.registerAdapter(RecordingAdapter()); //Register the adapter, essentially telling Hive how to read and write our Recording information into a box
+  await Hive.openBox<Recording>('recordings');
```
Should be run in the constructor for the Recording page, otherwise Hive has the `RecordingAdapter` registered for the entire application and if another page changes it the Recording page calls to Hive will fail.

pabloj2001 replied:
Let's remove this in favor of putting it in the Future function in the home page

pabloj2001 reviewed (changes requested):
Great changes! Left a few comments. We may also want to rethink how we do states, since we're setting empty states most of the time. We can talk about this further.

Alec-Hallman committed:
Tunning

Changes:
-fixed "durration" spelling error to duration
-Changed DisplayRecording to displayRecording to better show it's a method not a class
- removed void main from playback page as we won't be opening the app in it
-changed the controller listener code from the change state button press event to a initialize state method which runs on page creation.
- Made the Recording class do empty calls for now until I get the API up and running
-Changed the Emotions in the recording class to be a list (array) of items instead of just one item. This also involved creating a more specialized read and write method for Hive.

Note:
- There is no open or close box functionality just yet. I'm getting concerned on my time availability for this next week so while I do want to get it done. I also want to prioritize the API first. So I will be changing box open close soon. Just not this commit.

pabloj2001 commented on mobile/voice_journal_app/lib/RecordingPage.dart:
```diff
@@ -35,30 +46,38 @@ void displaySaved(){ //Create a little pop up letting the user know their reccor
+  recording = true; 
+  //rbox = await Hive.openBox<Recording>('recordings'); //open the box so we can put things inside of it.
+  rbox = Hive.box<Recording>('recordings');
+  currentRecording = Recording(id: '',title: formattedDateTime,audioFile: path,transcriptFile: '',timeStamp: presently,emotion:[Emotions.happy, Emotions.anger],isTranscribed: false,transcriptionId:'', duration: 0); //create the recording class
```
let's have emotions be an empty array at the beginning, we don't want to assume the emotions.

pabloj2001 commented on mobile/voice_journal_app/lib/RecordingPage.dart:
```diff
@@ -67,27 +86,36 @@ pauseRecording(){
+          _startCounting(); //Starts the displayed recording counter
+        }
+        setState(() { //change state (icon of the button should change)
+        });
```
instead of setting empty states, let's set `recording = true` for the new state instead of doing it in the `startRecording` function. Any variables used to render widgets in the `build` function should be set in the state.

pabloj2001 commented on mobile/voice_journal_app/lib/RecordingPage.dart:
```diff
@@ -170,8 +207,8 @@ class _RecordingPageState extends State<RecordingPage>{
-          splashColor: Colors.red[500],     //button click animation
+          backgroundColor: AppColors.accentColor, //button background colour
+          splashColor: Colors.red[200],     //button click animation
           child: rcrdIcon,
```
Instead of keeping a widget as a variable and updating it based on what we need, let's keep the widgets within the `build` function and show one or another based on some condition. For example:
```
child: recording ? const Icon(Icons.mic_off_outlined) : const Icon(Icons.pause),
```
Here we set one icon or another based on the recording variable that gets updated thru the state.

pabloj2001 commented on mobile/voice_journal_app/lib/home.dart:
```diff
@@ -1,77 +1,115 @@
 
-class HomePage extends StatelessWidget {
+//init variables
+bool boxOpened = false;
```
not used

pabloj2001 commented on mobile/voice_journal_app/lib/home.dart:
```diff
@@ -1,77 +1,115 @@
+                      color: AppColors.primaryColor, // Background color of the rectangle
+                      borderRadius: BorderRadius.circular(10), // Rounded corners
+                    ),
+                    child: ListView.builder( 
```
Let's set up a function that returns a `Future` and in that function we can open the box and return a list of recordings. Then here we can have the child be a `FutureBuilder` which will have the `ListView` builder as a child. This will allow us to load the recordings in the background, not having to await anything, and to open the box in that function instead of in the `Main` class. Take a look at FutureBuilders: https://www.educative.io/answers/what-is-a-futurebuilder

pabloj2001 replied:
Take a look at how Anika did it for the Stats page: https://github.com/LaurierCS/POD1/blob/895ca5886af919ce191a7d5d0decf0608cd65aac/mobile/voice_journal_app/lib/stats.dart

pabloj2001 reviewed (changes requested):
Great job! Left some comments on things we should address

pabloj2001 commented on mobile/voice_journal_app/lib/Emotions_enums.dart:
```diff
@@ -0,0 +1,10 @@
+  sad,
+  happy,
+  fear,
+  contempt,
```
We aren't using 'contempt' as one of our emotions so let's remove it here

pabloj2001 replied:
Just noticed these aren't in the same order as we specified in the schema, let's fix that so they're the same as the backend:
```
enum Emotions {
    happiness, // id = 0
    sadness, // id = 1
    anger, // id = 2
    fear, // id = 3
    disgust, // id = 4
    surprise, // id = 5
}
```

Alec-Hallman committed:
Pablo's Comments

- Corrected the Enum
- Hive Box now open and closes in the home page and recording page. (NOTE the register adapter is still in main as this is a one time set up operation. So keeping it in main seemed like a natural choice just to ensure it is set up once on app start up.)
- Added the future function to the homepage so that the list

pabloj2001 commented on mobile/voice_journal_app/lib/Playback.dart:
```diff
@@ -0,0 +1,154 @@
+    } else{ //if the recording is not playing
+      playbackRecording(); //play it
+    }
+    setState(() { //visual update for the pause and play button
```
Similarly to the recording page, let's update the `playing` variable in the state and let's change the button icon based on this variable instead of keeping the icon widget in a variable.

Alec-Hallman replied:
Oops! I should have done that proactively, 😰.

pabloj2001 commented on mobile/voice_journal_app/lib/home.dart:
```diff
@@ -1,77 +1,142 @@
+                      color: AppColors.primaryColor, // Background color of the rectangle
+                      borderRadius: BorderRadius.circular(10), // Rounded corners
                     ),
+                    child: RecordingList()
```
I like how you separated the recording list into a separate widget! Makes it so much cleaner and easier to read!

pabloj2001 reviewed (changes requested):
Good job making those changes! I haven't had a chance to run it but did want us to update the playback page to use the `playing` variable and the icons similarly to how we did in the recording page (thru the state)

Alec-Hallman committed:
Recording page tunning

-Removed the two empty change states under the event listener, and the pause play functions.
- Removed the play back icon variable

pabloj2001 reviewed (approved):
LGTM great job on these changes!
